src/mpid/ch4/src/ch4_coll.h

```diff
@@ -34,6 +34,12 @@ MPL_STATIC_INLINE_PREFIX int MPID_Barrier(MPIR_Comm * comm, MPIR_Errflag_t * err
             mpi_errno =
                 MPIDI_Barrier_intra_composition_beta(comm, errflag, ch4_algo_parameters_container);
             break;
+#ifdef MPIDI_BUILD_CH4_SHM
+        case MPIDI_CH4_Barrier_intra_composition_gamma_id:
```

Comment:
@AlexeyMalkhanov I'm having a little trouble picturing how this will work. Will the selection functions also need modification to actually use these compositions?

Reply:
@raffenet at the moment selection functions repeat legacy logic of collective selection. For the most collectives, there was no possibility to choose shm-specific algorithms. These shm-specific pieces in call phases can be treated as a way to handle shm algorithms which will be chosen based on decision tree parsing.

Reply:
@raffenet if you remember we previously introduced `is_single_node` attribute of a communicator for that purpose and based on the value of the attribute we chose netmod or shm inside a single composition which operates with the whole communicator. It turned out that the approach with `is_single_node` was dubious and we decided to introduce a separate composition for the case when the whole communicator resides inside one node.

Reply:
I see. I think I understand the concept, but without part 2 it's hard to completely visualize. Would it be much trouble to drop this patch until there is code ready to take advantage of it?